### ISSUE TITLE
updating (blockfrost) provider when changing network

### ISSUE THREAD

**Issue by GGAlanSmithee** [spacebudz/lucid#105]
Hi.

I am currently working on an example repo with a blockfrost proxy api, as discussed in https://github.com/spacebudz/lucid/issues/25#issuecomment-1145689280.

In this example, I want to be able to communicate to the blockforst API that corresponds to the wallet-active network. Ie. if the user - using Nami - switches network from Mainnet to Testnet, I want this to be reflected in the Blockfrost API.

This can be tracked with

```ts
window.cardano.nami.expermintal.on("networkChange", updateNetwork)
``` 

but my design-related question is how to best update the `Lucid` client. As I understand it, the only way to set the provider is at instantiation, ie.

```ts
const lucid = Lucid.new(new Blockfrost(url, key))
```

With this API, I could - using React - keep the lucid client as a state and override it, but that feels a bit awkward, since I would prefer it to be a global singletong of sorts.

Is it possible to simply update the provider, ie. `lucid.provider = newProvider` or is there an API for this?

Thanks.

**Comment by GGAlanSmithee** [comment]
I guess this goes for the second parameter - `network` - of `Lucid.new(blockfrostProvider, network` as well.

**Comment by alessandrokonrad** [comment]
Hey, so basically you want to have a function to update the provider from the existing Lucid instance?

**Comment by GGAlanSmithee** [comment]
Yes. But more directly, being able to react to switching networks. I have no deep understanding of lucid, so I don't know the inner workings of it, or what it entails, but just from an outside look, I guess both the `provider` and `network` arguments need to be reactive to this.

**Comment by GGAlanSmithee** [comment]
Either that Lucid automatically implicitly/internally updates (this is what I would expect but probably more complicated in terms of configuration) or explicitly through some API.

**Comment by GGAlanSmithee** [comment]
maybe something like

```ts
const lucid = Lucid.new({
  testnet: new Blockfrost(...),
  mainnet: new Blockfrost(...),
})
```

And in that case, the second param, `network`, would be dropped in favor of derriving that from the internal workings of the wallet integration. Again, don't knwo if this is technically possible to do.

**Comment by GGAlanSmithee** [comment]
Just for clarification, I think I can implement this in userland using `WalletApi.getNetworkId()`, which might be the correct way to go about it. I am in the process of creating a react hook which could incoorporate this, but an integrated, UI-library agnostic solution might be preferable.

**Comment by GGAlanSmithee** [comment]
Here is some triage code from a next.js page. It works, but does suffer from the design decisions I brought up earlier. Not really too pretty IMO. Also, it relies on the experimental event listener API, which might not be implemented by some wallets.

```tsx
import { useCallback, useEffect, useState } from "react"
import { isNil } from "lodash"
import { Blockfrost, Lucid, WalletApi } from "lucid-cardano"

const useWalletApi = () => {
  const [walletApi, setWalletApi] = useState<WalletApi>()

  useEffect(() => {
    window.cardano.nami.enable().then(setWalletApi)
  }, [])

  return walletApi
}

const useNetworkId = (walletApi?: WalletApi) => {
  const [networkId, setNetworkId] = useState<number>()

  const onNetworkChange = (newNetworkId: unknown) => {
    setNetworkId(newNetworkId as number)
  }

  useEffect(() => {
    if (!walletApi?.experimental) return

    walletApi.getNetworkId().then(setNetworkId)

    walletApi.experimental.on("networkChange", onNetworkChange)

    return () => {
      walletApi.experimental.off("networkChange", onNetworkChange)
    }
  }, [walletApi])

  return networkId
}

const useLucid = () => {
  const walletApi = useWalletApi()
  const networkId = useNetworkId(walletApi)

  const [lucid, setLucid] = useState<Lucid>()

  const createLucidInstance = useCallback(async () => {
    if (isNil(walletApi) || isNil(networkId)) return

    console.log("instantiating lucid instance with network id", networkId)

    const newLucidInstance = await Lucid.new(
      new Blockfrost(`/api/blockfrost/${networkId}`, ""),
      networkId === 0 ? "Testnet" : "Mainnet"
    )

    newLucidInstance.selectWallet(walletApi)

    return newLucidInstance
  }, [walletApi, networkId])

  useEffect(() => {
    createLucidInstance().then(setLucid)
  }, [walletApi, networkId, createLucidInstance])

  return {
    networkId,
    walletApi,
    lucid,
  }
}

const NetworkChangeTestPage = () => {
  const { networkId, walletApi, lucid } = useLucid()

  return (
    <div>
      <div>Network ID from wallet API: {networkId}</div>
    </div>
  )
}

export default NetworkChangeTestPage
```

**Comment by GGAlanSmithee** [comment]
Also, as a side note, it would be nice if the `Blockfrost` provider accepted `undefined` as protocolKey, with the assumption that this will be injected via a proxy or middleware of some sort.

**Comment by GGAlanSmithee** [comment]
@alessandrokonrad I created an example repo if you want to have a look at a potential user-land solution: https://github.com/GGAlanSmithee/cardano-lucid-blockfrost-proxy-example

**Comment by alessandrokonrad** [comment]
I think it's better right now to have a trigger event outside of Lucid. Nami is the only wallet that has such an event, but there could be also other criterias for changing a provider/network. It's more flexible to do it outside of Lucid, but I'm happy to provide a function that let's you update provider and network from an existing Lucid instance. Is that something useful to you?

> Also, as a side note, it would be nice if the Blockfrost provider accepted undefined as projectId, with the assumption that this will be injected via a proxy or middleware of some sort.

I think setting `projectId: ""` should be good enough. It could get a bit confusing otherwise.

**Comment by GGAlanSmithee** [comment]
> I'm happy to provide a function that let's you update provider and network from an existing Lucid instance. Is that something useful to you?

Yes please, because that way, I can keep the lucid instance as a singleton and also, if there are any internal side effects (now or in future versions) it would be easier for you to handle them.

> I think setting projectId: "" should be good enough. It could get a bit confusing otherwise.

Yeah, I understand. I think it can be confusing the way it is now as well, but I guess it depends on your use case. Either way, it's not a big issue!

**Comment by alessandrokonrad** [comment]
This should be resolved now with Lucid 0.7.1. There is a function `lucid.switchProvider()` and I made the projectId for the Blockfrost optional as well now :)

**Comment by GGAlanSmithee** [comment]
Aweome @alessandrokonrad I will update the example and after I've verified that it works, I'll close this issue!

**Comment by GGAlanSmithee** [comment]
@alessandrokonrad I've updated the example, and it works flawlessly. Just look at how much more beautiful and reasonable this code became: https://github.com/GGAlanSmithee/cardano-lucid-blockfrost-proxy-example/commit/87e380152b4fc3759c4f0307bad1077fa8489c03

Being able to keep lucid as a singleton will, IMO, allow for a much better DX than equivalents in the ETH web3 space. I am currently creating a react hook npm package for lucid (and some more QoL stuff) and I think this will allow me to work around some of the insance prop drilling that I've experienced when working with ETH web3 projects.

Cheers!
